modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -823,19 +823,26 @@ else if (prj.predicate() == F.<ClusterNode>alwaysTrue())
 
         long startTime = U.currentTimeMillis();
 
-        Map<UUID, Integer> top;
+        ServiceInfo desc;
 
         while (true) {
-            top = serviceTopology(name);
+             desc = lookupInRegisteredServices(name);
 
-            if (timeout == 0 || (top != null && !top.isEmpty()))
-                return top;
+            if (timeout == 0 && desc == null)
+                return null;
 
             synchronized (servicesTopsUpdateMux) {
-                long wait = timeout - (U.currentTimeMillis() - startTime);
+                if (desc != null && desc.topologyInitialized())
```

Comment:
This condition can be moved outside the synchronized block.

Reply:
I think it cannot be done that way.

If topology is updated before acquiring of servicesTopsUpdateMux but after this condiiton, we will be blocked until the next full msg.